### PR TITLE
Kodi

### PR DESCRIPTION
**Pull request by zguithues**
Kodi is the new name for XBMC.

These changes will rename the target and supporting files to kodi.  I also set it up to add the repository in order to get the latest version.

Ubuntu - Official TeamKodi PPA
Debian - Jessie Backports (recommended for debian in kodi wiki)
Kali - not supported.  will bail out of install with error.

I've tested with several releases, and it seems to work best with modern ubuntu (xenial+).

Features/Notes

- closes chroot when kodi is closed
- can play h.264 & h.265 out of the box (tested with Acer C720 peppy & Toshiba Chromebook 2 swanky)
- smb/samba works out of the box.  cannot scan, must enter port with hostname. 
- volume is separate from ChromeOS. (make sure it's turned up in chromeos)
- DOES NOT work with kiwi/extension
- I recommend making separate from other targets. (sudo sh ~/Downloads/crouton -t kodi -n kodi)

p.s. this is my first pull request ever, let me know if i did something wrong ;)

### CODE REVIEW

File: targets/kodi [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+# found in the LICENSE file.
+REQUIRES='x11'
+DESCRIPTION='Installs the KODI media player. (Approx. 140MB)'
+HOSTBIN='startkodi'
+CHROOTETC='kodi-keyboard.xml kodi-cycle.py'
+. "${TARGETSDIR:="$PWD"}/common"
+
+### Append to prepare.sh:
+if [ "$DISTRO" = 'debian' ]; then
+    echo "deb http://http.debian.net/debian jessie-backports main" | sudo tee /etc/apt/sources.list.d/kodi.list
```
Comment (dnschneid) [inline targets/kodi]: prepare already runs as root, so you can just use `>` to write the file.

File: targets/kodi [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+REQUIRES='x11'
+DESCRIPTION='Installs the KODI media player. (Approx. 140MB)'
+HOSTBIN='startkodi'
+CHROOTETC='kodi-keyboard.xml kodi-cycle.py'
+. "${TARGETSDIR:="$PWD"}/common"
+
+### Append to prepare.sh:
+if [ "$DISTRO" = 'debian' ]; then
```
Comment (dnschneid) [inline targets/kodi]: We shouldn't be using jessie-backports in sid, should we?  The right condition would then be `if release -le jessie; then`, which will apply only to debian releases jessie or earlier.

File: targets/kodi [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    echo "deb http://http.debian.net/debian jessie-backports main" | sudo tee /etc/apt/sources.list.d/kodi.list
+fi
+
+if [ "$DISTRO" = 'ubuntu' ]; then
+    sudo apt-get install -y software-properties-common
```
Comment (dnschneid) [inline targets/kodi]: you never need sudo in a target; you already have it

Comment (dnschneid) [inline targets/kodi]: Refer to this [old xfce target](https://github.com/dnschneid/crouton/blob/90ad47ff22e148db41ba627992ee5cc5d0c14175/targets/xfce#L13) for how to pull in PPAs.  We don't want to unnecessarily install things.

File: targets/kodi [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    sudo add-apt-repository -y ppa:team-xbmc/ppa
+fi
+
+if [ "$DISTRO" = 'kali' ]; then
+    error 99 "Kodi not supported on Kali Linux"
```
Comment (dnschneid) [inline targets/kodi]: This check should be done in the pre-prepare area, before you've gone through most of the crouton install.  See the ever-shrinking [unity target](https://github.com/dnschneid/crouton/blob/90ad47ff22e148db41ba627992ee5cc5d0c14175/targets/unity).

File: targets/kodi [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+fi
+
+sudo apt-get update
```
Comment (dnschneid) [inline targets/kodi]: This should only be run on distros that add the repository (e.g., put it in the if statements above).  Also, it seems apt-get update should be `|| true`'d in case the user has broken repos.

File: targets/kodi [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+# add a script in /etc/profile.d to link ~/.kodi/userdata/keymaps/keyboard.xml
+# to /etc/crouton/kodi-keyboard.xml
+
+profiledsh='/etc/profile.d/crouton-kodi-keymaps.sh'
+# Make sure symbolic link is setup on login
+echo '#!/bin/sh
+
+keyboardxmldir="$HOME/.kodi/userdata/keymaps"
+# Do not install if user is root, or $HOME does not exist
+if [ "`id -u`" -ne 0 -a -d "$HOME" -a ! -e "$keyboardxmldir/keyboard.xml" ]; then
```
Comment (dnschneid) [inline targets/kodi]: should be `"$(id -u)"`

File: targets/xbmc [hunk 7]
```diff
@@ -1,35 +0,0 @@
-#!/bin/sh -e
```
Comment (dnschneid) [inline targets/xbmc]: I don't think we can outright delete this target; we need to provide a migration path.

One quick-and-dirty solution would be to make xbmc a minimal target like [x11](https://github.com/dnschneid/crouton/blob/90ad47ff22e148db41ba627992ee5cc5d0c14175/targets/x11), with an unconditional `REQUIRES=kodi`, and add `PROVIDES='xbmc'` to the kodi target (like in [xorg](https://github.com/dnschneid/crouton/blob/90ad47ff22e148db41ba627992ee5cc5d0c14175/targets/xorg)).  Perhaps even leave DESCRIPTION blank and update installer/main.sh to avoid printing out any targets that have a blank DESCRIPTION for `-t list`.

The best solution would be to make xbmc a symlink to kodi, and then add some smarts to installer/main.sh to interpret a symlinked target as if it were the above.  This would be a little tricky (but not too hard) because the script would have to automatically populate PROVIDES by detecting any target symlinks that point to the selected target.

**Comment by zguithues** [comment]
Thanks for the suggestions!

I'll work through the list and resubmit when i've got them resolved.

File: targets/kodi (hunk 4 above)
Comment (zguithues) [inline targets/kodi]: I'm not certain what " if [ "${TARGETNOINSTALL:-c}" = 'c' ]; then " means..

Here is what i put at the top:

```
if [ "${TARGETNOINSTALL:-c}" = 'c' ]; then
    if [ "$DISTRO" = 'kali' ]; then
        error 99 "Kodi not supported on Kali Linux."
    fi
fi
```

Is this correct?

File: targets/xbmc (hunk 7 above)
Comment (zguithues) [inline targets/xbmc]: I've created the minimal target, but i'm unsure what exactly to add to installer/main.sh.

I assume it would be around line 306, but i'm not familiar with the syntax.

**Comment by zguithues** [comment]
Ok @dnschneid, I've made the requested changes.  I think everything is solid except for my questions regarding the kali linux and the xbmc migration stuff.

I haven't tested yet, as my chromebook is at home.  I will test it tonight and report back.

Thanks for your help getting this setup!  kodi is one of the main reasons i use crouton, and i'm glad to see it returning! :)

File: targets/xbmc [hunk 8]
```diff
@@ -2,34 +2,8 @@
 # Copyright (c) 2016 The crouton Authors. All rights reserved.
 # Use of this source code is governed by a BSD-style license that can be
 # found in the LICENSE file.
-REQUIRES='x11'
-DESCRIPTION='Installs the XBMC media player. (Approx. 140MB)'
-HOSTBIN='startxbmc'
-CHROOTETC='xbmc-keyboard.xml xbmc-cycle.py'
-. "${TARGETSDIR:="$PWD"}/common"
+REQUIRES='kodi'
+PROVIDES='xbmc'
```
Comment (dnschneid) [inline targets/xbmc]: PROVIDES would go in the kodi target, not the xbmc target.

File: targets/kodi (hunk 4 above)
Comment (dnschneid) [inline targets/kodi]: Yep!  It's not well-documented.

File: targets/xbmc (hunk 7 above)
Comment (dnschneid) [inline targets/xbmc]: It looks like in the help case, we set `TARGETS='help'`, on line 309 we set `TARGETNOINSTALL='y'` and then source the target file.  So it looks like the right tweak would be actually in [the target common file](https://github.com/dnschneid/crouton/blob/master/targets/common#L29), line 29, where it prints out the help text.  If DESCRIPTION is blank, we don't want to do that.

Comment (zguithues) [inline targets/xbmc]: i'm sorry, this went right over my head...  please dumb it down a bit ;)

I believe this is the only remaining item to resolve before this is good to merge.  Thanks a ton for your help! :)
Once merged, i'll reach out to the kodi guys and see if they'll feature it in their blog ;)

I also plan to make a wiki page that talks about it a bit.

Comment (dnschneid) [inline targets/xbmc]: Wrap [the common target](https://github.com/dnschneid/crouton/blob/master/targets/common#L30) lines 30 through 34 in a `if [ -n "$DESCRIPTION" ]` block :)

**Comment by zguithues** [comment]
ok @dnschneid I think i've resolved all of the issues.  I've tested installing to target xbmc and it installs kodi.  also, xbmc does not display in the target list due to it's blank description.

I tested on sid, and kodi was unable to start.  kodi in debian repositories is very old and i decided it would be better to just make it ubuntu only, as that is the only official repository.  I edited the description to: "Installs the KODI media player. Ubuntu Only. (Approx. 140MB)".

Let me know if you think i should handle anything differently.

**Comment by zguithues** [comment]
one thing i noticed... when i was testing, it would bail out for kali & debian before creating the chroot & downloading the bootstrap. but now it's downloading the whole thing before stopping. I'm not sure what i did to change it, and i can't duplicate the quick bailout.

Any thoughts? am i just crazy?

File: targets/kodi [hunk 9]
```diff
@@ -5,27 +5,19 @@
 if [ "${TARGETNOINSTALL:-c}" = 'c' ]; then
     if [ "$DISTRO" = 'kali' ]; then
         error 99 "Kodi not supported on Kali Linux."
+    elif [ "$DISTRO" = 'debian' ]; then
```
Comment (dnschneid) [inline targets/kodi]: Seems to be current in [sid](https://packages.debian.org/sid/video/kodi) (17), and jessie-backports seems to be version 16, which doesn't seem that old.  Can you get it to work so we don't artificially limit targets?

Generally we avoid creating targets that are only available in certain distros (unity being the obvious exception to that rule).

Comment (zguithues) [inline targets/kodi]: I'll play around with it...  When i tried to install it on sid, it wasn't able to connect to X and couldn't start kodi.  I'll see if i can figure out why tonight.

**Comment by zguithues** [comment]
@dnschneid 
I figured out why i was having trouble with kodi on kali & debian.  Its the xserver-xorg-legacy package!
detailed in #2875 

xfce is also broken in debian (and i assume kali) releases.

I'm sure there is a better way to solve this on the release script side, and i defer to your wisdom on this.  but the quickfix hammer i found was to add "remove xserver-xorg-legacy" to the target file.  Since it is such a detrimental package, and running that command doesn't have negative effects if the package is not installed; i just had it run for all releases

I'm running a few tests on various releases to see if i need to set anything specific.  I'll probably add back the jessie backports for jessie and older.  I'll update when i think it's ready to merge.

FYI, looking through the log, it looks like the package xserver-xorg-legacy was installed along with a bunch of xserver-xorg packages.  right after "Compiling /usr/local/lib/croutonfreon.so"
I'll add this to the open issue as well ;)

**Comment by zguithues** [comment]
turns out kodi doesn't work on Wheezy or Jessie.  I'm testing on trusty and precise right now.  I'll have it bail out on any releases that don't work.

should have final changes tomorrow

**Comment by zguithues** [comment]
I think i have everything resolved now @dnschneid 
Let me know if there is anything i missed :)

Thanks for all your help!

File: targets/xbmc (hunk 7 above)
Comment (zguithues) [inline targets/xbmc]: done

**Comment by zguithues** [comment]
I can't figure out how to clear the remaining requested changes...  I'm pretty sure I have addresses each requested change...

Should i clean the changes up into fewer commits and resubmit?

File: targets/kodi [hunk 10]
```diff
@@ -0,0 +1,77 @@
+#!/bin/sh -e
+# Copyright (c) 2016 The crouton Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+if [ "${TARGETNOINSTALL:-c}" = 'c' ]; then
+    if release -le wheezy; then
+        error 99 "Kodi not supported on Debian Wheezy or older."
+    fi
+    if release = precise; then
```
Comment (dnschneid) [inline targets/kodi]: better to use -eq here...this probably would work, though.

Comment (dnschneid) [inline targets/kodi]: this could be elif

File: targets/kodi [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        error 99 "Kodi not supported on Precise Ubuntu."
+    fi
+fi
+REQUIRES='x11'
+PROVIDES='xbmc'
+DESCRIPTION='Installs the KODI media player. Ubuntu Only. (Approx. 140MB)'
```
Comment (dnschneid) [inline targets/kodi]: "ubuntu only" isn't true anymore; you can just delete that

File: targets/kodi (hunk 10 above)
Comment (dnschneid) [inline targets/kodi]: (also down below)

File: targets/kodi [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    cat > '/etc/apt/sources.list.d/kodi.list' <<EOF
+# Bring in Jessie Backports for newer Kodi
+deb http://http.debian.net/debian jessie-backports main
+EOF
+    # Update database
+    apt-get update || true
+fi
+
+# Get rid of disruptive xserver-xorg-legacy package
+if release = stretch || release = sid || release = kali-rolling; then
```
Comment (dnschneid) [inline targets/kodi]: can be combined: `if release -ge stretch -eq kali-rolling`

File: targets/kodi [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    remove xserver-xorg-legacy
```
Comment (dnschneid) [inline targets/kodi]: wait, why is this workaround here, anyway?  It even happens before installing kodi.  Probably needs to be a separate pull request, implementing the workaround I mention [here](https://github.com/dnschneid/crouton/issues/2875#issuecomment-261073012).

Comment (zguithues) [inline targets/kodi]: I put this workaround in so it would work when testing.  I think a dummy package is probably the way to go.  I'll remove this workaround from the kodi pull request now.  I will try to take a crack at the dummy package later.

**Comment by zguithues** [comment]
OK, I've made the requested changes.

are we good to go now? :)

File: targets/kodi [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+DESCRIPTION='Installs the KODI media player. (Approx. 140MB)'
+HOSTBIN='startkodi'
+CHROOTETC='kodi-keyboard.xml kodi-cycle.py'
+. "${TARGETSDIR:="$PWD"}/common"
+
+### Append to prepare.sh:
+
+## Debian Overrides
+# Kodi wiki recommends using the Jessie Backports repository
+if release = jessie; then
```
Comment (dnschneid) [inline targets/kodi]: missed one :)

**Comment by dnschneid** [comment]
Also, I think the xbmc test needs to be changed into kodi: https://github.com/dnschneid/crouton/blob/master/test/tests/w6-xbmc

**Comment by zguithues** [comment]
ok, i changed the text to say kodi...  I'm not 100% sure what this is or does, so let me know if more is needed. :)

File: targets/kodi (hunk 14 above)
Comment (zguithues) [inline targets/kodi]: actually, there were 2 instances of "=", but now i'm not sure if i am supposed to switch both of them...

Comment (dnschneid) [inline targets/kodi]: yeah, always use -eq instead of = for the release command

File: targets/kodi [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+# Bring in Jessie Backports for newer Kodi
+deb http://http.debian.net/debian jessie-backports main
+EOF
+    # Update database
+    apt-get update || true
+fi
+
+## Ubuntu Overrides
+# Kodi has an official ppa for the latest stable version
+if [ "$DISTRO" -eq 'ubuntu' ]; then
```
Comment (dnschneid) [inline targets/kodi]: ah, sorry, but string comparisons should stick to ='s.

`release` is a special command that converts the release name into a number and does numerical comparisons.

Comment (zguithues) [inline targets/kodi]: i figured it was something like that...

**Comment by zguithues** [comment]
ok, that's all the pending changes requested...  anything else?

Thanks a ton for being so helpful! :)

**Comment by dnschneid** [comment]
Looks good to me!  Now I run some tests and merge if things pass!

Thanks for making and cleaning up the patches!

**Comment by dnschneid** [comment]
It looks like the PPA doesn't have any armhf/arm64 packages, so ARM devices will have to go without the PPA.  Currently, adding the PPA breaks install on ARM devices, since the metapackages are for any architecture and reference packages that cannot be fulfilled:
```
152 [i] The following packages have unmet dependencies:
153 [i]  kodi : Depends: kodi-bin (>= 2:17.3+git20170525.0741-final-0xenial) but 15.2+dfsg1-3ubuntu1.1 is to be installed
153 [i]         Depends: libshairplay0 but it is not installable
153 [i]         Depends: libcec4 but it is not installable
153 [i]         Recommends: libvdpau1 but it is not going to be installed
153 [i]         Recommends: i965-va-driver but it is not installable or
153 [i]                     libva-intel-vaapi-driver but it is not installable
```

**Comment by zguithues** [comment]
I'll set it to not add the PPA for armhf/arm64 devices.
It looks like old 15.2 kodi is in the default repositories, should i let it install that, or should i just bail out if arm/arm64?  I don't have a way to test how it works on armhf/arm64.

**Comment by zguithues** [comment]
ok, i set it to only add the ppa on x86,  can you test to see if it installs, or if i should just block arm? @dnschneid

**Comment by dnschneid** [comment]
I'll give it a shot.  Thanks!

**Comment by zguithues** [comment]
@dnschneid i've been using kodi and the one annoyance that i still have is the screen timeout...

reading the page on the wiki here: https://github.com/dnschneid/crouton/wiki/Power-manager-overrides
it seems like i should setup the wrapper script, but i feel like this could/should be implemented through the startkodi executable.  Either way, I'm not really sure how i should proceed.  Is there a precedent that I should follow, or do you have some insight?

Regarding which settings to adjust.  I believe we want to disable the dim_ms & off_ms for both plugged and unplugged.  probably need to adjust the suspend_ms as well...

Alternatively, is there a way for kodi to report that it's being "active" when playing video, so that it doesn't trigger any timeouts?

**Comment by zguithues** [comment]
I created the Kodi wiki page here: https://github.com/dnschneid/crouton/wiki/Kodi-Media-Center

I'll fluff it out in the next few weeks.

File: host-bin/startkodi [hunk 16]
```diff
@@ -9,10 +9,10 @@ APPLICATION="${0##*/}"
...
-Wraps enter-chroot to start an XBMC session.
+Wraps enter-chroot to start an KODI session.
 By default, it will log into the primary user on the first chroot found.
 
 Options are directly passed to enter-chroot; run enter-chroot to list them."
 
-exec sh -e "`dirname "\`readlink -f -- "$0"\`"`/enter-chroot" -t xbmc "$@" "" \
-    exec xinit /usr/bin/xbmc
+exec sh -e "`dirname "\`readlink -f -- "$0"\`"`/enter-chroot" -t kodi "$@" "" \
+    exec xinit /usr/bin/kodi --standalone
```
Comment (dnschneid) [inline host-bin/startkodi]: `exec croutonpowerd -i xinit /usr/bin/kodi --standalone` might do what you want with respect to power management.  It will prevent the system from idly going to sleep as long as the command (xinit /usr/bin/kodi in this case) is running.

Comment (zguithues) [reply host-bin/startkodi]: Thanks a ton!  I'll check it out!

**Comment by zguithues** [comment]
I have made the change that disables power management while kodi is running.  I believe this was my last remaining issue to resolve.

How has your testing gone @dnschneid ?  Do you think it's ready to merge with master?

**Comment by dnschneid** [comment]
Looks good from a testing point of view, although we need to fix the xserver-xorg-legacy thing.  Thanks again for this patch!

**Comment by zguithues** [comment]
Thanks so much for your help @dnschneid !

I'll keep it updated if i find anything that needs fixed.
Next step is to tell the kodi people they just got a new platform ;)

**Comment by DennisLfromGA** [comment]
RE: https://github.com/dnschneid/crouton/issues/3395

Not sure exactly what device this is since `edit-chroot -all` wasn't run but it's an ARM according to the `chrome` target output. They specified the `xbmc` target and got "E: Unable to locate package kodi" :(

**Comment by zguithues** [comment]
Thanks for pointing it out to me, i responded to the issue.  I think the problem is that he was using trusty, and kodi isn't being maintained on arm trusty.
